File: windows/PokeFolio.Desktop/Backend/PokeFolioApiPayloads.cs
using System.Net;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace PokeFolio.Desktop.Backend;

internal static class PokeFolioApiPayloads
{
    private const int MaximumCatalogRequestBytes = 16 * 1024;
    private const int MaximumSyncRequestBytes = 1024 * 1024;
    private static readonly UTF8Encoding StrictUtf8 = new(false, true);
    private static readonly Dictionary<string, string> CatalogProviderTcgs =
        new(StringComparer.OrdinalIgnoreCase)
        {
            ["pokemon-tcg-api"] = "pokemon",
            ["tcgdex"] = "pokemon",
            ["ygoprodeck"] = "yugioh",
            ["optcgapi"] = "onepiece"
        };
    private static readonly HashSet<string> CatalogReferenceProperties =
        new(StringComparer.Ordinal)
        {
            "provider",
            "providerCardId",
            "tcg",
            "name",
            "setCode",
            "number"
        };
    private static readonly HashSet<string> DeviceProperties =
        new(StringComparer.Ordinal)
        {
            "id",
            "name",
            "platform",
            "createdAt",
            "lastSeenAt",
            "current"
        };
    private static readonly JsonSerializerOptions JsonOptions = new(JsonSerializerDefaults.Web)
    {
        PropertyNameCaseInsensitive = false,
        UnmappedMemberHandling = JsonUnmappedMemberHandling.Disallow
    };

    public static byte[] SerializeLogin(
        string email,
        string password,
        string deviceName)
    {
        ValidateLoginInput(email, password, deviceName);
        return JsonSerializer.SerializeToUtf8Bytes(new
        {
            email = email.Trim(),
            password,
            deviceName = deviceName.Trim(),
            platform = "windows"
        }, JsonOptions);
    }

    public static byte[] SerializePasswordResetRequest(string email)
    {
        ValidateEmail(email);
        return JsonSerializer.SerializeToUtf8Bytes(new
        {
            email = email.Trim()
        }, JsonOptions);
    }

    public static byte[] SerializePasswordChange(
        string currentPassword,
        string newPassword)
    {
        ValidatePassword(currentPassword, 1, nameof(currentPassword), "Current password");
        ValidatePassword(newPassword, 12, nameof(newPassword), "New password");
        if (string.Equals(currentPassword, newPassword, StringComparison.Ordinal))
        {
            throw new ArgumentException(
                "New password must differ from the current password.",
                nameof(newPassword));
        }
        return JsonSerializer.SerializeToUtf8Bytes(new
        {
            currentPassword,
            newPassword
        }, JsonOptions);
    }

    public static byte[] SerializePasswordResetConfirm(
        string email,
        string token,
        string newPassword)
    {
        ValidateEmail(email);
        if (token is null || token.Length is < 32 or > 512)
        {
            throw new ArgumentException(
                "Password reset token must contain 32 to 512 characters.",
                nameof(token));
        }
        if (newPassword is null || newPassword.Length is < 12 or > 128)
        {
            throw new ArgumentException(
                "New password must contain 12 to 128 characters.",
                nameof(newPassword));
        }
        return JsonSerializer.SerializeToUtf8Bytes(new
        {
            email = email.Trim(),
            token,
            newPassword
        }, JsonOptions);
    }

    public static byte[] SerializeRefresh(string refreshToken) =>
        JsonSerializer.SerializeToUtf8Bytes(new { refreshToken }, JsonOptions);

    public static byte[] NormalizeCatalogCardReference(string json)
    {
        ArgumentNullException.ThrowIfNull(json);
        byte[] input = StrictUtf8.GetBytes(json);
        try
        {
            if (input.Length is < 2 or > MaximumCatalogRequestBytes)
            {
                throw new ArgumentException(
                    "Catalog card reference has an invalid size.",
                    nameof(json));
            }

            using JsonDocument document = JsonDocument.Parse(input);
            JsonElement root = document.RootElement;
            if (root.ValueKind != JsonValueKind.Object ||
                HasDuplicateProperty(root) ||
                root.EnumerateObject().Count() != CatalogReferenceProperties.Count ||
                root.EnumerateObject().Any(property =>
                    !CatalogReferenceProperties.Contains(property.Name)))
            {
                throw new ArgumentException(
                    "Catalog card reference must contain exactly the contract fields.",
                    nameof(json));
            }

            string provider = RequiredString(root, "provider").Trim().ToLowerInvariant();
            string providerCardId = RequiredString(root, "providerCardId")
                .Trim()
                .ToLowerInvariant();
            string tcg = RequiredString(root, "tcg").Trim().ToLowerInvariant();
            string name = NormalizeCatalogText(root, "name", 240);
            string setCode = NormalizeCatalogText(root, "setCode", 64);
            string number = NormalizeCatalogText(root, "number", 64);

            if (!CatalogProviderTcgs.TryGetValue(provider, out string? providerTcg))
            {
                throw new ArgumentException("Catalog provider is not supported.", nameof(json));
            }
            if (!string.Equals(providerTcg, tcg, StringComparison.Ordinal))
            {
                throw new ArgumentException(
                    "Catalog provider does not match the selected TCG.",
                    nameof(json));
            }
            if (providerCardId.Length is < 1 or > 160 ||
                providerCardId.Any(character =>
                    !char.IsAsciiLetterOrDigit(character) &&
                    character is not '-' and not '_' and not '.' and not ':' and not '/'))
            {
                throw new ArgumentException(
                    "Provider card id contains invalid characters.",
                    nameof(json));
            }

            return JsonSerializer.SerializeToUtf8Bytes(new
            {
                provider,
                providerCardId,
                tcg,
                name,
                setCode,
                number
            }, JsonOptions);
        }
        catch (JsonException error)
        {
            throw new ArgumentException(
                "Catalog card reference must be valid JSON.",
                nameof(json),
                error);
        }
        finally
        {
            CryptographicOperations.ZeroMemory(input);
        }
    }

    public static byte[] ValidateSyncOperationBatch(string json)
    {
        ArgumentNullException.ThrowIfNull(json);
        byte[] body = StrictUtf8.GetBytes(json);
        try
        {
            if (body.Length is < 2 or > MaximumSyncRequestBytes)
            {
                throw new ArgumentException("Sync operation batch has an invalid size.", nameof(json));
            }
            using JsonDocument document = JsonDocument.Parse(body);
            if (document.RootElement.ValueKind != JsonValueKind.Object ||
                HasDuplicateProperty(document.RootElement))
            {
                throw new ArgumentException(
                    "Sync operation batch must be one JSON object without duplicate properties.",
                    nameof(json));
            }
            return body;
        }
        catch (JsonException error)
        {
            CryptographicOperations.ZeroMemory(body);
            throw new ArgumentException("Sync operation batch must be valid JSON.", nameof(json), error);
        }
        catch
        {
            CryptographicOperations.ZeroMemory(body);
            throw;
        }
    }

    public static PokeFolioAuthSessionEnvelope ParseSession(byte[] body)
    {
        PokeFolioAuthSessionEnvelope? envelope;
        try
        {
            using JsonDocument document = JsonDocument.Parse(body);
            if (document.RootElement.ValueKind != JsonValueKind.Object ||
                HasDuplicateProperty(document.RootElement))
            {
                throw new JsonException(
                    "Authentication response must be one JSON object without duplicate properties.");
            }
            envelope = document.RootElement.Deserialize<PokeFolioAuthSessionEnvelope>(JsonOptions);
        }
        catch (Exception error) when (error is JsonException or NotSupportedException)
        {
            throw new InvalidDataException("Authentication response is invalid.", error);
        }
        if (envelope is null ||
            envelope.UserId == Guid.Empty ||
            string.IsNullOrEmpty(envelope.AccessToken) ||
            envelope.AccessToken.Length is < 32 or > 16_384 ||
            string.IsNullOrEmpty(envelope.RefreshToken) ||
            envelope.RefreshToken.Length is < 32 or > 1024 ||
            envelope.AccessTokenExpiresAt == default ||
            envelope.Device is null ||
            envelope.Device.Id == Guid.Empty ||
            string.IsNullOrWhiteSpace(envelope.Device.Name) ||
            envelope.Device.Name.Length > 120 ||
            !string.Equals(envelope.Device.Platform, "windows", StringComparison.Ordinal) ||
            envelope.Device.CreatedAt == default ||
            envelope.Device.LastSeenAt == default ||
            !envelope.Device.Current)
        {
            throw new InvalidDataException("Authentication response violates the client contract.");
        }
        return envelope;
    }

    public static PokeFolioApiProblem ParseProblem(PokeFolioRawResponse response)
    {
        string fallbackCode = response.Status == (int)HttpStatusCode.Unauthorized
            ? "authentication_required"
            : "api_request_failed";
        string fallbackTitle = $"Backend request failed with HTTP {response.Status}.";
        if (response.Body.Length == 0)
        {
            return new PokeFolioApiProblem(response.Status, fallbackCode, fallbackTitle);
        }

        try
        {
            using JsonDocument document = JsonDocument.Parse(response.Body);
            JsonElement root = document.RootElement;
            if (root.ValueKind != JsonValueKind.Object)
            {
                return new PokeFolioApiProblem(response.Status, fallbackCode, fallbackTitle);
            }

            string code = TryReadString(root, "code") ?? fallbackCode;
            string title = TryReadString(root, "title") ?? fallbackTitle;
            IReadOnlyDictionary<string, string[]>? errors = TryReadErrors(root);
            return new PokeFolioApiProblem(response.Status, code, title, errors);
        }
        catch (JsonException)
        {
            return new PokeFolioApiProblem(response.Status, fallbackCode, fallbackTitle);
        }
    }

    public static IReadOnlyList<PokeFolioManagedDevice> ParseDeviceList(
        string json,
        Guid expectedCurrentDeviceId)
    {
        if (expectedCurrentDeviceId == Guid.Empty)
        {
            throw new InvalidDataException("Current device id is missing.");
        }
        byte[] body;
        try
        {
            body = StrictUtf8.GetBytes(json ?? throw new InvalidDataException(
                "Device response is missing."));
        }
        catch (EncoderFallbackException error)
        {
            throw new InvalidDataException("Device response is not valid UTF-8.", error);
        }
        try
        {
            using JsonDocument document = JsonDocument.Parse(body);
            JsonElement root = document.RootElement;
            if (root.ValueKind != JsonValueKind.Array || HasDuplicateProperty(root))
            {
                throw new InvalidDataException("Device response must be one strict JSON array.");
            }
            JsonElement.ArrayEnumerator entries = root.EnumerateArray();
            var result = new List<PokeFolioManagedDevice>();
            int currentCount = 0;
            foreach (JsonElement entry in entries)
            {
                if (result.Count >= 1000 ||
                    entry.ValueKind != JsonValueKind.Object ||
                    entry.EnumerateObject().Count() != DeviceProperties.Count ||
                    entry.EnumerateObject().Any(property =>
                        !DeviceProperties.Contains(property.Name)))
                {
                    throw new InvalidDataException("Device response violates the API contract.");
                }
                if (!entry.TryGetProperty("id", out JsonElement idValue) ||
                    !idValue.TryGetGuid(out Guid id) || id == Guid.Empty ||
                    !entry.TryGetProperty("createdAt", out JsonElement createdValue) ||
                    !createdValue.TryGetDateTimeOffset(out DateTimeOffset createdAt) ||
                    createdAt == default ||
                    !entry.TryGetProperty("lastSeenAt", out JsonElement seenValue) ||
                    !seenValue.TryGetDateTimeOffset(out DateTimeOffset lastSeenAt) ||
                    lastSeenAt == default ||
                    !entry.TryGetProperty("current", out JsonElement currentValue) ||
                    currentValue.ValueKind is not (JsonValueKind.True or JsonValueKind.False))
                {
                    throw new InvalidDataException("Device response contains invalid values.");
                }
                string name = DeviceString(entry, "name", 120);
                string platform = DeviceString(entry, "platform", 32);
                bool current = currentValue.GetBoolean();
                if ((current && id != expectedCurrentDeviceId) ||
                    (!current && id == expectedCurrentDeviceId))
                {
                    throw new InvalidDataException("Device response has an invalid current marker.");
                }
                if (current) currentCount++;
                result.Add(new PokeFolioManagedDevice(
                    id, name, platform, createdAt, lastSeenAt, current));
            }
            if (currentCount != 1)
            {
                throw new InvalidDataException(
                    "Device response must contain the current session once.");
            }
            return result;
        }
        catch (JsonException error)
        {
            throw new InvalidDataException("Device response is invalid JSON.", error);
        }
        finally
        {
            CryptographicOperations.ZeroMemory(body);
        }
    }

    public static string DecodeBody(byte[] body)
    {
        try
        {
            return StrictUtf8.GetString(body);
        }
        catch (DecoderFallbackException error)
        {
            throw new InvalidDataException("Backend response is not valid UTF-8.", error);
        }
    }

    private static bool HasDuplicateProperty(JsonElement value)
    {
        if (value.ValueKind == JsonValueKind.Array)
        {
            return value.EnumerateArray().Any(HasDuplicateProperty);
        }
        if (value.ValueKind != JsonValueKind.Object) return false;

        var propertyNames = new HashSet<string>(StringComparer.Ordinal);
        foreach (JsonProperty property in value.EnumerateObject())
        {
            if (!propertyNames.Add(property.Name) || HasDuplicateProperty(property.Value))
            {
                return true;
            }
        }
        return false;
    }

    private static string? TryReadString(JsonElement root, string propertyName) =>
        root.TryGetProperty(propertyName, out JsonElement value) &&
        value.ValueKind == JsonValueKind.String
            ? value.GetString()
            : null;

    private static string RequiredString(JsonElement root, string propertyName)
    {
        if (!root.TryGetProperty(propertyName, out JsonElement value) ||
            value.ValueKind != JsonValueKind.String)
        {
            throw new ArgumentException(
                $"Catalog field {propertyName} must be a string.",
                nameof(root));
        }
        return value.GetString()!;
    }

    private static string DeviceString(
        JsonElement root,
        string propertyName,
        int maximumLength)
    {
        if (!root.TryGetProperty(propertyName, out JsonElement value) ||
            value.ValueKind != JsonValueKind.String)
        {
            throw new InvalidDataException($"Device field {propertyName} must be text.");
        }
        string result = value.GetString()!;
        if (string.IsNullOrWhiteSpace(result) ||
            result.Length > maximumLength)
        {
            throw new InvalidDataException($"Device field {propertyName} is invalid.");
        }
        return result;
    }

    private static string NormalizeCatalogText(
        JsonElement root,
        string propertyName,
        int maximumLength)
    {
        string value = RequiredString(root, propertyName).Trim();
        if (value.Length is < 1 ||
            value.Length > maximumLength ||
            value.Any(char.IsControl))
        {
            throw new ArgumentException(
                $"Catalog field {propertyName} is invalid.",
                nameof(root));
        }
        return value;
    }

    private static IReadOnlyDictionary<string, string[]>? TryReadErrors(JsonElement root)
    {
        if (!root.TryGetProperty("errors", out JsonElement errorsElement) ||
            errorsElement.ValueKind != JsonValueKind.Object)
        {
            return null;
        }

        var errors = new Dictionary<string, string[]>(StringComparer.Ordinal);
        foreach (JsonProperty property in errorsElement.EnumerateObject())
        {
            if (property.Value.ValueKind != JsonValueKind.Array) continue;
            string[] messages = property.Value.EnumerateArray()
                .Where(item => item.ValueKind == JsonValueKind.String)
                .Select(item => item.GetString()!)
                .ToArray();
            if (messages.Length > 0) errors[property.Name] = messages;
        }
        return errors.Count == 0 ? null : errors;
    }

    private static void ValidateLoginInput(string email, string password, string deviceName)
    {
        ValidateEmail(email);
        if (string.IsNullOrEmpty(password) || password.Length > 128)
        {
            throw new ArgumentException("Password must contain 1 to 128 characters.", nameof(password));
        }
        if (string.IsNullOrWhiteSpace(deviceName) || deviceName.Trim().Length > 120)
        {
            throw new ArgumentException("Device name must contain 1 to 120 characters.", nameof(deviceName));
        }
    }

    private static void ValidateEmail(string email)
    {
        if (string.IsNullOrWhiteSpace(email) || email.Trim().Length > 254)
        {
            throw new ArgumentException("Email must contain 1 to 254 characters.", nameof(email));
        }
    }

    private static void ValidatePassword(
        string password,
        int minimumLength,
        string parameterName,
        string fieldName)
    {
        if (password is null || password.Length < minimumLength || password.Length > 128)
        {
            throw new ArgumentException(
                $"{fieldName} must contain {minimumLength} to 128 characters.",
                parameterName);
        }
    }
}

internal sealed record PokeFolioAuthSessionEnvelope(
    Guid UserId,
    string AccessToken,
    string RefreshToken,
    DateTimeOffset AccessTokenExpiresAt,
    PokeFolioAuthDeviceEnvelope Device);

internal sealed record PokeFolioAuthDeviceEnvelope(
    Guid Id,
    string Name,
    string Platform,
    DateTimeOffset CreatedAt,
    DateTimeOffset LastSeenAt,
    bool Current);

internal sealed record PokeFolioManagedDevice(
    Guid Id,
    string Name,
    string Platform,
    DateTimeOffset CreatedAt,
    DateTimeOffset LastSeenAt,
    bool Current);

File: app/src/main/java/de/pokefolio/app/backend/PokeFolioApiPayloads.java
package de.pokefolio.app.backend;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.CharBuffer;
import java.nio.charset.CharacterCodingException;
import java.nio.charset.CodingErrorAction;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.time.OffsetDateTime;
import java.time.format.DateTimeParseException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.UUID;

final class PokeFolioApiPayloads {
    private static final int MAXIMUM_CATALOG_REQUEST_BYTES = 16 * 1024;
    private static final int MAXIMUM_SYNC_REQUEST_BYTES = 1024 * 1024;
    private static final UUID EMPTY_UUID = new UUID(0L, 0L);
    private static final Map<String, String> CATALOG_PROVIDER_TCGS;
    private static final Set<String> CATALOG_REFERENCE_PROPERTIES = Collections.unmodifiableSet(
            new HashSet<>(Arrays.asList(
                    "provider", "providerCardId", "tcg", "name", "setCode", "number")));
    private static final Set<String> SESSION_PROPERTIES = Collections.unmodifiableSet(
            new HashSet<>(Arrays.asList(
                    "userId", "accessToken", "refreshToken", "accessTokenExpiresAt", "device")));
    private static final Set<String> DEVICE_PROPERTIES = Collections.unmodifiableSet(
            new HashSet<>(Arrays.asList(
                    "id", "name", "platform", "createdAt", "lastSeenAt", "current")));

    static {
        Map<String, String> providers = new LinkedHashMap<>();
        providers.put("pokemon-tcg-api", "pokemon");
        providers.put("tcgdex", "pokemon");
        providers.put("ygoprodeck", "yugioh");
        providers.put("optcgapi", "onepiece");
        CATALOG_PROVIDER_TCGS = Collections.unmodifiableMap(providers);
    }

    private PokeFolioApiPayloads() {
    }

    static byte[] serializeLogin(String email, String password, String deviceName) {
        validateLoginInput(email, password, deviceName);
        try {
            JSONObject json = new JSONObject();
            json.put("email", email.trim());
            json.put("password", password);
            json.put("deviceName", deviceName.trim());
            json.put("platform", "android");
            return json.toString().getBytes(StandardCharsets.UTF_8);
        } catch (JSONException error) {
            throw new IllegalArgumentException("Account request cannot be serialized.", error);
        }
    }

    static byte[] serializeRefresh(String refreshToken) {
        if (refreshToken == null || refreshToken.length() < 32 || refreshToken.length() > 1024) {
            throw new IllegalArgumentException("Refresh token has an invalid length.");
        }
        try {
            return new JSONObject()
                    .put("refreshToken", refreshToken)
                    .toString()
                    .getBytes(StandardCharsets.UTF_8);
        } catch (JSONException error) {
            throw new IllegalArgumentException("Refresh request cannot be serialized.", error);
        }
    }

    static byte[] serializePasswordResetRequest(String email) {
        validateEmail(email);
        try {
            return new JSONObject()
                    .put("email", email.trim())
                    .toString()
                    .getBytes(StandardCharsets.UTF_8);
        } catch (JSONException error) {
            throw new IllegalArgumentException(
                    "Password reset request cannot be serialized.", error);
        }
    }

    static byte[] serializePasswordChange(String currentPassword, String newPassword) {
        validatePassword(currentPassword, 1, "Current password");
        validatePassword(newPassword, 12, "New password");
        if (currentPassword.equals(newPassword)) {
            throw new IllegalArgumentException(
                    "New password must differ from the current password.");
        }
        try {
            return new JSONObject()
                    .put("currentPassword", currentPassword)
                    .put("newPassword", newPassword)
                    .toString()
                    .getBytes(StandardCharsets.UTF_8);
        } catch (JSONException error) {
            throw new IllegalArgumentException(
                    "Password change request cannot be serialized.", error);
        }
    }

    static byte[] serializePasswordResetConfirm(
            String email,
            String token,
            String newPassword
    ) {
        validateEmail(email);
        if (token == null || token.length() < 32 || token.length() > 512) {
            throw new IllegalArgumentException(
                    "Password reset token must contain 32 to 512 characters.");
        }
        if (newPassword == null || newPassword.length() < 12 || newPassword.length() > 128) {
            throw new IllegalArgumentException(
                    "New password must contain 12 to 128 characters.");
        }
        try {
            return new JSONObject()
                    .put("email", email.trim())
                    .put("token", token)
                    .put("newPassword", newPassword)
                    .toString()
                    .getBytes(StandardCharsets.UTF_8);
        } catch (JSONException error) {
            throw new IllegalArgumentException(
                    "Password reset confirmation cannot be serialized.", error);
        }
    }

    static byte[] normalizeCatalogCardReference(String json) {
        byte[] input = encodeRequest(json, MAXIMUM_CATALOG_REQUEST_BYTES, "Catalog card reference");
        try {
            StrictJsonValidator.validateObject(json);
            JSONObject root = new JSONObject(json);
            requireExactProperties(root, CATALOG_REFERENCE_PROPERTIES, "catalog card reference");
            String provider = requiredRequestString(root, "provider").trim().toLowerCase(java.util.Locale.ROOT);
            String providerCardId = requiredRequestString(root, "providerCardId")
                    .trim()
                    .toLowerCase(java.util.Locale.ROOT);
            String tcg = requiredRequestString(root, "tcg").trim().toLowerCase(java.util.Locale.ROOT);
            String name = normalizeCatalogText(root, "name", 240);
            String setCode = normalizeCatalogText(root, "setCode", 64);
            String number = normalizeCatalogText(root, "number", 64);
            String expectedTcg = CATALOG_PROVIDER_TCGS.get(provider);
            if (expectedTcg == null) {
                throw new IllegalArgumentException("Catalog provider is not supported.");
            }
            if (!expectedTcg.equals(tcg)) {
                throw new IllegalArgumentException("Catalog provider does not match the selected TCG.");
            }
            if (providerCardId.length() < 1 || providerCardId.length() > 160
                    || !isProviderCardId(providerCardId)) {
                throw new IllegalArgumentException("Provider card id contains invalid characters.");
            }
            JSONObject normalized = new JSONObject();
            normalized.put("provider", provider);
            normalized.put("providerCardId", providerCardId);
            normalized.put("tcg", tcg);
            normalized.put("name", name);
            normalized.put("setCode", setCode);
            normalized.put("number", number);
            return normalized.toString().getBytes(StandardCharsets.UTF_8);
        } catch (IOException | JSONException error) {
            throw new IllegalArgumentException("Catalog card reference must be valid contract JSON.", error);
        } finally {
            Arrays.fill(input, (byte) 0);
        }
    }

    static byte[] validateSyncOperationBatch(String json) {
        byte[] body = encodeRequest(json, MAXIMUM_SYNC_REQUEST_BYTES, "Sync operation batch");
        try {
            StrictJsonValidator.validateObject(json);
            return body;
        } catch (IOException error) {
            Arrays.fill(body, (byte) 0);
            throw new IllegalArgumentException(
                    "Sync operation batch must be one JSON object without duplicate properties.",
                    error);
        }
    }

    static AuthEnvelope parseSession(byte[] body) throws IOException {
        String json = decodeUtf8(body);
        StrictJsonValidator.validateObject(json);
        try {
            JSONObject root = new JSONObject(json);
            requireExactProperties(root, SESSION_PROPERTIES, "authentication response");
            UUID userId = requiredUuid(root, "userId");
            String accessToken = requiredString(root, "accessToken", 32, 16_384);
            String refreshToken = requiredString(root, "refreshToken", 32, 1024);
            Instant accessTokenExpiresAt = requiredInstant(root, "accessTokenExpiresAt");
            Object rawDevice = root.opt("device");
            if (!(rawDevice instanceof JSONObject)) {
                throw new IOException("Authentication response device must be an object.");
            }
            JSONObject device = (JSONObject) rawDevice;
            requireExactProperties(device, DEVICE_PROPERTIES, "authentication device");
            UUID deviceId = requiredUuid(device, "id");
            String deviceName = requiredString(device, "name", 1, 120);
            String platform = requiredString(device, "platform", 1, 32);
            if (!"android".equals(platform)) {
                throw new IOException("Authentication response belongs to another platform.");
            }
            Instant createdAt = requiredInstant(device, "createdAt");
            Instant lastSeenAt = requiredInstant(device, "lastSeenAt");
            if (!(device.opt("current") instanceof Boolean)
                    || !device.optBoolean("current", false)) {
                throw new IOException("Authentication response device is not current.");
            }
            return new AuthEnvelope(
                    userId,
                    accessToken,
                    refreshToken,
                    accessTokenExpiresAt,
                    new DeviceEnvelope(deviceId, deviceName, platform, createdAt, lastSeenAt));
        } catch (JSONException error) {
            throw new IOException("Authentication response is invalid JSON.", error);
        }
    }

    static PokeFolioApiProblem parseProblem(int status, byte[] body) {
        String fallbackCode = status == 401 ? "authentication_required" : "api_request_failed";
        String fallbackTitle = "Backend request failed with HTTP " + status + ".";
        if (body == null || body.length == 0) {
            return new PokeFolioApiProblem(status, fallbackCode, fallbackTitle, null);
        }
        try {
            String json = decodeUtf8(body);
            StrictJsonValidator.validateObject(json);
            JSONObject root = new JSONObject(json);
            String code = optionalString(root, "code", 1, 120, fallbackCode);
            String title = optionalString(root, "title", 1, 500, fallbackTitle);
            Map<String, List<String>> errors = parseErrors(root.opt("errors"));
            return new PokeFolioApiProblem(status, code, title, errors);
        } catch (Exception ignored) {
            return new PokeFolioApiProblem(status, fallbackCode, fallbackTitle, null);
        }
    }

    static JSONArray parseDeviceList(String json, UUID expectedCurrentDeviceId)
            throws IOException {
        if (expectedCurrentDeviceId == null || EMPTY_UUID.equals(expectedCurrentDeviceId)) {
            throw new IOException("Current device id is missing.");
        }
        StrictJsonValidator.validateArray(json);
        try {
            JSONArray input = new JSONArray(json);
            if (input.length() > 1000) {
                throw new IOException("Device response contains too many entries.");
            }
            JSONArray normalized = new JSONArray();
            int currentCount = 0;
            for (int index = 0; index < input.length(); index++) {
                Object raw = input.opt(index);
                if (!(raw instanceof JSONObject)) {
                    throw new IOException("Device response entries must be objects.");
                }
                JSONObject device = (JSONObject) raw;
                requireExactProperties(device, DEVICE_PROPERTIES, "device response");
                UUID id = requiredUuid(device, "id");
                String name = requiredString(device, "name", 1, 120);
                String platform = requiredString(device, "platform", 1, 32);
                Instant createdAt = requiredInstant(device, "createdAt");
                Instant lastSeenAt = requiredInstant(device, "lastSeenAt");
                Object rawCurrent = device.opt("current");
                if (!(rawCurrent instanceof Boolean)) {
                    throw new IOException("current must be a boolean.");
                }
                boolean current = (Boolean) rawCurrent;
                if (current) {
                    currentCount++;
                    if (!expectedCurrentDeviceId.equals(id)) {
                        throw new IOException("Device response marks another session as current.");
                    }
                } else if (expectedCurrentDeviceId.equals(id)) {
                    throw new IOException("Current device is not marked as current.");
                }
                normalized.put(new JSONObject()
                        .put("id", id.toString())
                        .put("name", name)
                        .put("platform", platform)
                        .put("createdAt", createdAt.toString())
                        .put("lastSeenAt", lastSeenAt.toString())
                        .put("current", current));
            }
            if (currentCount != 1) {
                throw new IOException("Device response must contain the current session once.");
            }
            return normalized;
        } catch (JSONException error) {
            throw new IOException("Device response is invalid JSON.", error);
        }
    }

    static String decodeUtf8(byte[] body) throws IOException {
        if (body == null) return "";
        try {
            return StandardCharsets.UTF_8.newDecoder()
                    .onMalformedInput(CodingErrorAction.REPORT)
                    .onUnmappableCharacter(CodingErrorAction.REPORT)
                    .decode(ByteBuffer.wrap(body))
                    .toString();
        } catch (CharacterCodingException error) {
            throw new IOException("Backend response is not valid UTF-8.", error);
        }
    }

    private static byte[] encodeRequest(String value, int maximumBytes, String label) {
        if (value == null) throw new IllegalArgumentException(label + " is required.");
        try {
            ByteBuffer encoded = StandardCharsets.UTF_8.newEncoder()
                    .onMalformedInput(CodingErrorAction.REPORT)
                    .onUnmappableCharacter(CodingErrorAction.REPORT)
                    .encode(CharBuffer.wrap(value));
            int size = encoded.remaining();
            if (size < 2 || size > maximumBytes) {
                throw new IllegalArgumentException(label + " has an invalid size.");
            }
            byte[] result = new byte[size];
            encoded.get(result);
            return result;
        } catch (CharacterCodingException error) {
            throw new IllegalArgumentException(label + " is not valid Unicode.", error);
        }
    }

    private static String requiredRequestString(JSONObject value, String property)
            throws JSONException {
        Object raw = value.opt(property);
        if (!(raw instanceof String)) {
            throw new IllegalArgumentException("Catalog field " + property + " must be text.");
        }
        return (String) raw;
    }

    private static String normalizeCatalogText(
            JSONObject value,
            String property,
            int maximumLength
    ) throws JSONException {
        String normalized = requiredRequestString(value, property).trim();
        if (normalized.isEmpty() || normalized.length() > maximumLength
                || containsControlCharacter(normalized)) {
            throw new IllegalArgumentException("Catalog field " + property + " is invalid.");
        }
        return normalized;
    }

    private static boolean containsControlCharacter(String value) {
        for (int index = 0; index < value.length(); index++) {
            if (Character.isISOControl(value.charAt(index))) return true;
        }
        return false;
    }

    private static boolean isProviderCardId(String value) {
        for (int index = 0; index < value.length(); index++) {
            char character = value.charAt(index);
            if ((character >= 'a' && character <= 'z')
                    || (character >= '0' && character <= '9')
                    || character == '-' || character == '_' || character == '.'
                    || character == ':' || character == '/') {
                continue;
            }
            return false;
        }
        return true;
    }

    private static Map<String, List<String>> parseErrors(Object raw) throws JSONException {
        if (!(raw instanceof JSONObject)) return null;
        JSONObject value = (JSONObject) raw;
        Map<String, List<String>> errors = new LinkedHashMap<>();
        Iterator<String> properties = value.keys();
        while (properties.hasNext()) {
            String property = properties.next();
            Object rawMessages = value.opt(property);
            if (!(rawMessages instanceof JSONArray) || property.length() > 120) continue;
            JSONArray messages = (JSONArray) rawMessages;
            if (messages.length() > 50) continue;
            List<String> accepted = new ArrayList<>();
            for (int index = 0; index < messages.length(); index++) {
                Object message = messages.opt(index);
                if (message instanceof String && !((String) message).trim().isEmpty()
                        && ((String) message).length() <= 500) {
                    accepted.add((String) message);
                }
            }
            if (!accepted.isEmpty()) errors.put(property, Collections.unmodifiableList(accepted));
        }
        return errors.isEmpty() ? null : errors;
    }

    private static void requireExactProperties(
            JSONObject value,
            Set<String> expected,
            String label
    ) throws IOException {
        Set<String> actual = new HashSet<>();
        Iterator<String> properties = value.keys();
        while (properties.hasNext()) actual.add(properties.next());
        if (!actual.equals(expected)) {
            throw new IOException(label + " does not match the API contract.");
        }
    }

    private static UUID requiredUuid(JSONObject value, String property) throws IOException {
        String text = requiredString(value, property, 36, 36);
        try {
            UUID id = UUID.fromString(text);
            if (EMPTY_UUID.equals(id) || !id.toString().equalsIgnoreCase(text)) {
                throw new IllegalArgumentException("UUID is not canonical.");
            }
            return id;
        } catch (IllegalArgumentException error) {
            throw new IOException(property + " must be a non-empty canonical UUID.", error);
        }
    }

    private static Instant requiredInstant(JSONObject value, String property) throws IOException {
        String text = requiredString(value, property, 1, 80);
        try {
            return OffsetDateTime.parse(text).toInstant();
        } catch (DateTimeParseException error) {
            throw new IOException(property + " must be an ISO timestamp.", error);
        }
    }

    private static String requiredString(
            JSONObject value,
            String property,
            int minimum,
            int maximum
    ) throws IOException {
        Object raw = value.opt(property);
        if (!(raw instanceof String)) throw new IOException(property + " must be text.");
        String text = (String) raw;
        if (text.length() < minimum || text.length() > maximum || text.trim().isEmpty()) {
            throw new IOException(property + " has an invalid length.");
        }
        return text;
    }

    private static String optionalString(
            JSONObject value,
            String property,
            int minimum,
            int maximum,
            String fallback
    ) {
        Object raw = value.opt(property);
        if (!(raw instanceof String)) return fallback;
        String text = (String) raw;
        return text.length() >= minimum && text.length() <= maximum && !text.trim().isEmpty()
                ? text : fallback;
    }

    private static void validateLoginInput(String email, String password, String deviceName) {
        validateEmail(email);
        if (password == null || password.isEmpty() || password.length() > 128) {
            throw new IllegalArgumentException("Password must contain 1 to 128 characters.");
        }
        if (deviceName == null || deviceName.trim().isEmpty() || deviceName.trim().length() > 120) {
            throw new IllegalArgumentException("Device name must contain 1 to 120 characters.");
        }
    }

    private static void validateEmail(String email) {
        if (email == null || email.trim().isEmpty() || email.trim().length() > 254) {
            throw new IllegalArgumentException("Email must contain 1 to 254 characters.");
        }
    }

    private static void validatePassword(
            String password,
            int minimumLength,
            String fieldName
    ) {
        if (password == null
                || password.length() < minimumLength
                || password.length() > 128) {
            throw new IllegalArgumentException(
                    fieldName + " must contain " + minimumLength + " to 128 characters.");
        }
    }

    static final class AuthEnvelope {
        final UUID userId;
        final String accessToken;
        final String refreshToken;
        final Instant accessTokenExpiresAt;
        final DeviceEnvelope device;

        AuthEnvelope(
                UUID userId,
                String accessToken,
                String refreshToken,
                Instant accessTokenExpiresAt,
                DeviceEnvelope device
        ) {
            this.userId = userId;
            this.accessToken = accessToken;
            this.refreshToken = refreshToken;
            this.accessTokenExpiresAt = accessTokenExpiresAt;
            this.device = device;
        }
    }

    static final class DeviceEnvelope {
        final UUID id;
        final String name;
        final String platform;
        final Instant createdAt;
        final Instant lastSeenAt;

        DeviceEnvelope(
                UUID id,
                String name,
                String platform,
                Instant createdAt,
                Instant lastSeenAt
        ) {
            this.id = id;
            this.name = name;
            this.platform = platform;
            this.createdAt = createdAt;
            this.lastSeenAt = lastSeenAt;
        }
    }
}
